### ISSUE TITLE
Evaluation metrics - Regression / Generative models

### ISSUE THREAD

**Issue by csudre** [Project-MONAI/MONAI#1127]
**Is your feature request related to a problem? Please describe.**
A list of the metrics to be implemented to be applied in the case of a regression/generative task with possible links for implementation / reference

- Mean absolute error [https://pytorch.org/docs/master/generated/torch.nn.L1Loss.html#torch.nn.L1Loss](url)
- Mean squared error [https://pytorch.org/docs/master/generated/torch.nn.MSELoss.html#torch.nn.MSELoss](url)
- Smooth L1 Loss [https://pytorch.org/docs/master/generated/torch.nn.SmoothL1Loss.html#torch.nn.SmoothL1Loss](url)
- Root MSE [](url)
- Peak Signal to Noise Ratio [https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/image_ops_impl.py#L3358](url)
- Structural Similarity Index Measure [https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/image_ops_impl.py#L3569](url)
- Multi-scale SSIM [https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/ops/image_ops_impl.py#L3648](url
- Negative Guassian Log Likelihood [https://pytorch.org/docs/stable/nn.html#torch.nn.NLLLoss](url)
- Maximum Mean Discrepancy [https://github.com/ZongxianLee/MMD_Loss.Pytorch/blob/master/mmd_loss.py#L5](url)
- Inception Score [https://github.com/sbarratt/inception-score-pytorch](url)
- Frechet Inception Distance [https://github.com/xuqiantong/GAN-Metrics](url)

**Additional context**
This comes from the work on metrics by task type done by the Metrics Task Force

**Comment by masadcv** [comment]
Hi @wyli and @Nic-Ma , Is this task open for contribution? 
If no one is working on it, I would like to take up this task to cover regression metrics in the first instance. Can you also indicate what type of support is required for these metrics?

**Comment by Nic-Ma** [comment]
Hi @masadcv ,

These items are proposed by our metrics working group, open for contribution.
Welcome your contribution PR if you are interested in some of them.

Thanks in advance.

**Comment by masadcv** [comment]
Many thanks @Nic-Ma  - I will start looking into this towards the end of this week. I will have a proposal for Regression metrics in the first instance.

**Comment by danieltudosiu** [comment]
@masadcv did you have time to check the MS-SSIM? I implemented a prototype of the the 4-G-MS-SSIM [1] by extending pytorch_msssim [2], and was wondering if we could merge our progress for this ticket. 

[1] Li, C. and Bovik, A.C., 2010. Content-partitioned structural similarity index for image quality assessment. Signal Processing: Image Communication, 25(7), pp.517-526.
[2] https://pypi.org/project/pytorch-msssim/

**Comment by masadcv** [comment]
@danieltudosiu I have not started on this yet, got busy in other stuff. It would definitely be worth including as much as we can. My idea for this was to go for regression metrics such as: MSE, MAE, SSIM, PSNR, etc and it seems the MS SSIM may be a good fit. As these are just evaluation metrics, I dont see a point in putting a lot of effort in making them differentiable in the first instance, unless there is a need to.

I would appreciate your help with this. Are you happy to contribute to SSIM, MS SSIM and 4-G-MS-SSIM as they all seem related? I will update once I have some working prototype for proposal for including these in monai. A ref implementation of SSIM is available through skimage as well: https://scikit-image.org/docs/dev/auto_examples/transform/plot_ssim.html

**Comment by masadcv** [comment]
@danieltudosiu I have added the proposal implementation. Can you have a look and see if your code can be merged into this? Please note this is still an early work in progress, so there may be changes/improvements that may be done in coming days to it. To include your implementation, you may find RegressionMetric class useful where I have added some basic shape checking and reduction functionality. 

@Nic-Ma, @wyli It would be very helpful if you can have an initial look at my proposal for adding regression metrics. Please do indicate if some aspects need a different approach, I can work on it. This is currently missing SSIM and related metrics for regression.

**Comment by danieltudosiu** [comment]
@masadcv @wyli @Nic-Ma Shouldn't all our metrics be DDP compliant similar to those: 

https://pytorch.org/ignite/_modules/ignite/metrics/mean_squared_error.html

**Comment by Nic-Ma** [comment]
Hi @danieltudosiu ,

Yes, all the metrics should be DDP compliant in MONAI, I think all the existing metrics support DDP.

Thanks.

**Comment by danieltudosiu** [comment]
Update:
Privately I have a working implementation of all the known SSIM derivatives:
- SSIM, G-SSIM, 3-SSIM, 4-SSIM
- MS-SSIM, MS-G-SSIM, 3-MS-G-SSIM, 4-MS-G-SSIM

At the moment it has an instability issue where due to padding it can result in NaN. After discussions with my supervisor I will shift to cropping the gradient maps to fit the SSIM map so all the measurements are in areas where the convolutions are valid.

**Comment by wyli** [comment]
#2084 partly addressed the issue

**Comment by masadcv** [comment]
Update: 
Here is a simple regression network (with image denoising task) to demonstrate the usage of Regression Metrics:
https://colab.research.google.com/drive/1Ueuay4xvNkF7io1hGxah7a9Xr2J4zY6L?usp=sharing

**Comment by danieltudosiu** [comment]
Update:

I have all SSIM but the star [1] enhancement done, it has some issues with the masking and the authors of the paper can't seem to be reachable. Discussing with a Spanish researcher about possible bugs.

But it is very memory intensive, with more than 12GB VRAM required for a single 222^3 image. I am talking with an Nvidia Engineer to try and optimize it a bit more. 

Also, does anybody know how you can code a differentiable binary mask? I've done some tricks but they might be unstable for backward propagation.

[1] Rouse, D.M. and Hemami, S.S., 2008, February. Analyzing the role of visual structure in the recognition of natural image content with multi-scale SSIM. In Human Vision and Electronic Imaging XIII (Vol. 6806, p. 680615). International Society for Optics and Photonics.
